RFC 310: Add config encryption to ExternalServiceStore

arussellsaw opened the pull request:
closes: #18059 #18058 

This PR updates the database.ExternalServiceStore to encrypt & decrypt the config field in and out of the database. This is a compromise as i think we'd rather only encrypt/decrypt last minute, but our config architecture makes this hard to do. eventually we'd like to move to a new format where we can encrypt individual fields, and decrypt them on access. whilst this PR doesn't achieve all of our goals, it does incrementally improve the security of our external service storage.

#### In defense of a package level variable:
Whilst i agree that package level state is bad, the keyring.Default doesn't necessarily contain any state, but is a set of functions defined by config. if our architecture were at the point where there was a centralised point to inject dependencies, i would definitely opt to inject this into the store, but considering these stores are created ad-hoc all over the place, the scope of this change would be huge, and require threading these dependencies through a huge number of unrelated packages. I think we should aim to migrate to a better injection pattern in the long term, but be pragmatic about how we write code at the moment. I think in this case using dependency injection for this keyring would only stand to introduce more package coupling, and make things worse rather than better.

<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->

asdine commented on cmd/frontend/internal/cli/config.go:
```diff
@@ -94,10 +80,24 @@ func handleConfigOverrides() error {
 
-	if overrideGlobalSettings != "" {
-		globalSettingsBytes, err := ioutil.ReadFile(overrideGlobalSettings)
+func overrideGlobalSettings(ctx context.Context) error {
```
nit/ Since we are rewriting this function, I think we should give it the db handler and replace calls to `database.GlobalSettings` by `database.Settings(db)`

arussellsaw replied:
good call, updated

asdine reviewed (approved):
LGTM

arussellsaw commented:
i've made a bunch of changes based on some feedback from @tsenart:
* replaced config_encrypted with encryption_key_id as i imagine there will be a time where we'll want to know this.
* added some actual validation of encryption to my test cases, this caught some bugs
* made keyring.Default update with config.Watch
i haven't added test cases for the config refactor yet as it seems like it'll be quite hard to test. i think i'll work out some manual test steps before i merge this, and validate changes in k8s.sgdev.org

tsenart commented on internal/database/external_services.go:
```diff
@@ -525,16 +533,63 @@ func (e *ExternalServiceStore) Create(ctx context.Context, confGet func() *conf.
-
 	es.Unrestricted = !gjson.GetBytes(normalized, "authorization").Exists()
 
+	config, configWasEncrypted, err := e.maybeEncryptConfig(ctx, es.Config)
```
This variable name `configWasEncrypted` no longer reflects what it's in it. How about `keyID`?

tsenart commented on internal/encryption/cloudkms/cloud_kms.go:
```diff
@@ -31,6 +32,18 @@ type Key struct {
 	client *kms.KeyManagementClient
 }
 
+func (k *Key) Identifier(ctx context.Context) (string, error) {
```
Why not just `ID`?

tsenart replied:
Also, would it be cleaner to call `GetCryptoKey` in the `NewKey` constructor and store the result in the struct? Then here we wouldn't need to pass a context and there would be no error.

arussellsaw replied:
i guess there's more of a chance the key rotates between those two calls? right now there's small chance so not a huge difference. right now the default rotation interval is 30d

tsenart commented on internal/encryption/keyring/ring.go:
```diff
@@ -3,15 +3,57 @@ package keyring
+		}
+		newRing, err := NewRing(ctx, newConfig)
+		if err != nil {
+			log15.Error("creating encryption keyring", "error", err)
```
Is logging an error enough? Won't everything break? Perhaps a panic is more appropriate?

arussellsaw replied:
you mentioned config validation in the call we had, do you know where i should look for that?

tsenart replied:
https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/site.go#L154

tsenart commented on internal/encryption/keyring/ring.go:
```diff
@@ -3,15 +3,57 @@ package keyring
 )
 
+var (
+	mu          sync.Mutex
```
Since this will be mostly read, I think we can use a RWMutex to reduce contention?

tsenart commented on internal/encryption/migrate/migrate.go:
```diff
@@ -0,0 +1,24 @@
+	"github.com/sourcegraph/sourcegraph/internal/encryption"
+)
+
+type MigratorKey struct {
```
Godoc please.

arussellsaw replied:
oh this shouldn't be committed, my bad

tsenart reviewed (approved):
Assuming a good manual test plan, LGTM

arussellsaw commented:
i ran a bunch of manual tests editing config in both the site admin UI & config file, external service config file, & external services settings. I think i've found a bug where it won't detect two separate edits to site config whilst running, but that was preexisting.

unknwon commented on internal/encryption/keyring/ring.go:
```diff
@@ -3,15 +3,64 @@ package keyring
+)
+
+func Default() Ring {
+	mu.RLock()
+	defer mu.RUnlock()
+	return defaultRing
```
Just noting another pattern we use is `atomic.Value`. For example, https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/globals/globals.go#L22-26

arussellsaw replied:
this field is write-very-rarely and read often, i don't think atomic.Value is worth the compexity overhead here. generally i stay away from atomic unless it's absolutely proven that i need it.
